prov/sockets/src/sock_ep_msg.c

```diff
@@ -379,440 +378,196 @@ static int sock_ep_cm_getpeer(struct fid_ep *ep, void *addr, size_t *addrlen)
 	return (len == sizeof(struct sockaddr_in)) ? 0 : -FI_ETOOSMALL;
 }
 
-static int sock_ep_cm_create_socket(void)
+static inline int sock_cm_send(int fd, const void *buf, int len)
 {
-	int sock, optval;
-	sock = socket(AF_INET, SOCK_DGRAM, 0);
-	if (sock < 0)
-		return 0;
-
-	optval = 1;
-	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
-		       &optval, sizeof(optval)))
-		SOCK_LOG_ERROR("setsockopt failed\n");
-	return sock;
-}
-
-static int sock_ep_cm_enqueue_msg(struct sock_cm_entry *cm,
-				  const struct sockaddr_in *addr,
-				  void *msg, size_t len,
-				  fid_t fid, struct sock_eq *eq)
-{
-	char c = 0;
-	int ret = 0;
-	struct sock_cm_msg_list_entry *list_entry;
-
-	list_entry = calloc(1, sizeof(*list_entry) + len);
-	if (!list_entry)
-		return -FI_ENOMEM;
-
-	list_entry->msg_len = len;
-	memcpy(&list_entry->msg[0], msg, len);
-	memcpy(&list_entry->addr, addr, sizeof(*addr));
-	list_entry->fid = fid;
-	list_entry->eq = eq;
-
-	fastlock_acquire(&cm->lock);
-	dlist_insert_tail(&list_entry->entry, &cm->msg_list);
-	fastlock_release(&cm->lock);
-
-	ret = ofi_write_socket(cm->signal_fds[0], &c, 1);
-	if (ret != 1) {
-		SOCK_LOG_DBG("failed to signal\n");
-		ret = -FI_EIO;
-	} else {
-		ret = 0;
-		SOCK_LOG_DBG("Enqueued CM Msg\n");
-	}
-	return ret;
-}
-
-static int sock_ep_cm_send_msg(struct sock_cm_entry *cm,
-			       const struct sockaddr_in *addr,
-			       void *msg, size_t len)
-{
-	int ret;
-#if ENABLE_DEBUG
-	char sa_ip[INET_ADDRSTRLEN] = {0};
-	memcpy(sa_ip, inet_ntoa(addr->sin_addr), INET_ADDRSTRLEN);
-#endif
-	ret = sendto(cm->sock, (char *) msg, len, 0,
-		     (struct sockaddr *) addr, sizeof(*addr));
-	SOCK_LOG_DBG("Sending %d to %s:%d\n", ret, sa_ip, ntohs(addr->sin_port));
-	return (ret == len) ? 0 : -1;
-}
-
-static void sock_ep_cm_release_entry(struct sock_cm_msg_list_entry *msg_entry)
-{
-	struct sock_conn_hdr *msg_hdr;
-	struct fi_eq_cm_entry cm_entry;
-	struct sock_ep *sock_ep;
-
-	msg_hdr = (struct sock_conn_hdr *) msg_entry->msg;
-	if (msg_hdr->type == SOCK_CONN_SHUTDOWN) {
-		memset(&cm_entry, 0, sizeof(cm_entry));
-		cm_entry.fid = msg_entry->fid;
-		sock_ep = container_of(cm_entry.fid, struct sock_ep, ep.fid);
-		if (!sock_ep->attr->cm.shutdown_received) {
-			sock_ep->attr->cm.shutdown_received = 1;
-			sock_ep_disable(&sock_ep->ep);
-			if (sock_eq_report_event(msg_entry->eq, FI_SHUTDOWN, &cm_entry,
-			     sizeof(cm_entry), 0))
-				SOCK_LOG_ERROR("Error in writing to EQ\n");
+	int ret, done = 0;
+	while (done != len) {
+		ret = ofi_write_socket(fd, buf + done, len - done);
```

Comment:
this code failed to compile on windows. could you change it to:
`ret = ofi_write_socket(fd, (const char*)buf + done, len - done);`
thank you